rln/src/merkle.rs

```diff
@@ -1,26 +1,36 @@
 // Adapted from https://github.com/kilic/rln/blob/master/src/merkle.rs
 //
 use crate::poseidon::{Poseidon as Hasher, PoseidonParams};
-// TODO Replace these with arkworks
-use sapling_crypto::bellman::pairing::ff::{Field, PrimeField, PrimeFieldRepr};
-use sapling_crypto::bellman::pairing::Engine;
+use ark_ff::{Field, PrimeField};
+//use ff::{Field, PrimeField, PrimeFieldRepr};
+use ark_ec::PairingEngine;
+
 use std::io::{self, Error, ErrorKind};
 use std::{collections::HashMap, hash::Hash};
+use std::str::FromStr;
+use std::fmt::Debug;
+
+use ark_bn254::Fr;
 
 pub struct IncrementalMerkleTree<E>
 where
-    E: Engine,
+    E: PairingEngine,
 {
     pub current_index: usize,
     merkle_tree: MerkleTree<E>,
 }
 
+// XXX: I don't understand this second where clause
 impl<E> IncrementalMerkleTree<E>
 where
-    E: Engine,
+    E: PairingEngine,
+    <<E as PairingEngine>::Fr as FromStr>::Err: Debug,
```

Comment:
`Fr` and `Err` are _associated types_ to `PairingEngine` and `FromStr` traits: https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html

the `as` terminology enables you to refer to said associated types.
At the end of the day, what this clause is trying to do is that `Err` is `Debug` which can then enable to debug print any `Err` in this `impl` block.

The `Err` we are talking about is the `Err` type associated to the `FromStr` trait so it has to do `< ..as FromStr>::Err` to specify that we are binding `FromStr::Err`. And then again to specify that we are binding `PairingEngine::Fr`.